commit df335618ed87eecdef44a95e453e345a55a14ad8
Author: Amit Kapila <akapila@postgresql.org>
Date:   Thu Jul 24 03:51:55 2025 +0000

    Fix cfbot failure caused by commit 228c370868.
    
    Ensure the test waits for the apply worker to exit after disabling the
    subscription. This is necessary to safely enable the retain_dead_tuples
    option. Also added a similar wait in another part of the test to prevent
    unintended apply worker activity that could lead to test failures
    post-subscription disable.
    
    Reported by Michael Paquier as per cfbot.
    
    Author: Zhijie Hou <houzj.fnst@fujitsu.com>
    Discussion: https://postgr.es/m/aIGLgfRJIBwExoPj@paquier.xyz

diff --git a/src/test/subscription/t/035_conflicts.pl b/src/test/subscription/t/035_conflicts.pl
index 7458d7fba7e..976d53a870e 100644
--- a/src/test/subscription/t/035_conflicts.pl
+++ b/src/test/subscription/t/035_conflicts.pl
@@ -226,10 +226,15 @@ ok( $stderr =~
 	"altering retain_dead_tuples is not allowed for enabled subscription");
 
 # Disable the subscription
 $node_A->psql('postgres', "ALTER SUBSCRIPTION $subname_AB DISABLE;");
 
+# Wait for the apply worker to stop
+$node_A->poll_query_until('postgres',
+	"SELECT count(*) = 0 FROM pg_stat_activity WHERE backend_type = 'logical replication apply worker'"
+);
+
 # Enable retain_dead_tuples for disabled subscription
 ($cmdret, $stdout, $stderr) = $node_A->psql('postgres',
 	"ALTER SUBSCRIPTION $subname_AB SET (retain_dead_tuples = true);");
 ok( $stderr =~
 	  /NOTICE:  deleted rows to detect conflicts would not be removed until the subscription is enabled/,
@@ -276,10 +281,15 @@ is($result, qq(1|1
 2|2), 'check replicated insert on node B');
 
 # Disable the logical replication from node B to node A
 $node_A->safe_psql('postgres', "ALTER SUBSCRIPTION $subname_AB DISABLE");
 
+# Wait for the apply worker to stop
+$node_A->poll_query_until('postgres',
+	"SELECT count(*) = 0 FROM pg_stat_activity WHERE backend_type = 'logical replication apply worker'"
+);
+
 $node_B->safe_psql('postgres', "UPDATE tab SET b = 3 WHERE a = 1;");
 $node_A->safe_psql('postgres', "DELETE FROM tab WHERE a = 1;");
 
 ($cmdret, $stdout, $stderr) = $node_A->psql(
 	'postgres', qq(VACUUM (verbose) public.tab;)
